File: Algorithms/SelectionSort.go
package main

import "fmt"

// [12, 5, 10], n = 3
// iter 1: currIndex = 0, minIndex = 0; lookAhead = 1
// 		1 < 3 <TRUE>
//			5 < 12 <TRUE>
//				minIndex = 1
//			lookAhead = 2
//		2 < 3 <TRUE>
//			10 < 5 <FALSE>
//			lookAhead = 3
// 		3 < 3 <FALSE>
//		[5, 12, 10]
// ..
// ..
// [5, 10, 12]

// Iterates over an array while putting the smallest values into their correct place.
// Time complexity: O(n^2)
// Space complexity: O(1)
// Uses: The good thing about selection sort is it never makes more than O(n)
// swaps and can be useful when memory write is a costly operation.
func selectionSort(arr []int) {
	n := len(arr)

	// iterate over the array
	for currIndex := 0; currIndex < n-1; currIndex++ {
		// index of the minimum item
		minIndex := currIndex
		// look ahead
		lookAhead := currIndex + 1

		// loop while our look ahead is within bounds
		for lookAhead < n {
			// check if our look ahead value is less than our minimum index value
			if arr[lookAhead] < arr[minIndex] {
				// update minimum index
				minIndex = lookAhead
			}
			// increment our look ahead
			lookAhead++
		}

		// swap our minimum index value and our current index value
		arr[minIndex], arr[currIndex] = arr[currIndex], arr[minIndex]
	}
}

func main() {
	arr := []int{12, 5, 10}

	selectionSort(arr)

	fmt.Println(arr)
}

File: Algorithms/SelectionSort.js
function SelectionSort(arr) {
    // if the array is less than 2 elements, it's already sorted
    if (arr.length < 2) return arr;

    // loop over the array
    for (let i = 0; i < arr.length; i++) {
        // index of the minimum element going through this loop
        let minIdx = i;
        // lookahead to check and see for potential smaller indexes
        let lookahead = minIdx + 1;

        // while lookahead hasn't gotten to the end, increment it
        while (lookahead < arr.length) {
            // if lookahead's element is smaller
            if (arr[lookahead] < arr[minIdx]) {
                // update our min index
                minIdx = lookahead
            }

            lookahead++
        }

        // swap our current element with the minimum element we found
        const temp = arr[i];
        arr[i] = arr[minIdx];
        arr[minIdx] = temp;
    }

    return arr
}

const arr = [5, 2, 1, 6, 13, -3];

console.log(SelectionSort(arr))

